impeller/entity/contents/atlas_contents.cc

```diff
@@ -57,6 +64,96 @@ void AtlasContents::SetCullRect(std::optional<Rect> cull_rect) {
   cull_rect_ = cull_rect;
 }
 
+struct AtlasBlenderKey {
+  Color color;
+  Rect rect;
+  int32_t color_key;
+
+  static int32_t ToColorKey(Color color) {
+    return (((std::lround(color.alpha * 255) & 0xff) << 24) |
+            ((std::lround(color.red * 255) & 0xff) << 16) |
+            ((std::lround(color.green * 255) & 0xff) << 8) |
+            ((std::lround(color.blue * 255) & 0xff) << 0)) &
+           0xFFFFFFFF;
+  }
+
+  struct Hash {
+    std::size_t operator()(const AtlasBlenderKey& key) const {
+      return fml::HashCombine(key.color_key, key.rect.size.width,
+                              key.rect.size.height, key.rect.origin.x,
+                              key.rect.origin.y);
+    }
+  };
+
+  struct Equal {
+    bool operator()(const AtlasBlenderKey& lhs,
+                    const AtlasBlenderKey& rhs) const {
+      return lhs.rect == rhs.rect && lhs.color_key == rhs.color_key;
+    }
+  };
+};
+
+std::shared_ptr<SubAtlasResult> AtlasContents::GenerateSubAtlas() const {
+  FML_DCHECK(colors_.size() > 0 && blend_mode_ != BlendMode::kSource &&
+             blend_mode_ != BlendMode::kDestination);
+
+  std::unordered_map<AtlasBlenderKey, std::vector<Matrix>,
+                     AtlasBlenderKey::Hash, AtlasBlenderKey::Equal>
+      sub_atlas = {};
+
+  for (auto i = 0u; i < texture_coords_.size(); i++) {
+    AtlasBlenderKey key = {
+        .color = colors_[i],
+        .rect = texture_coords_[i],
+        .color_key = AtlasBlenderKey::ToColorKey(colors_[i])};
+    if (sub_atlas.find(key) == sub_atlas.end()) {
+      sub_atlas[key] = {transforms_[i]};
+    } else {
+      sub_atlas[key].push_back(transforms_[i]);
+    }
+  }
+
+  auto result = std::make_shared<SubAtlasResult>();
+  Scalar x_offset = 0.0;
+  Scalar y_offset = 0.0;
+  Scalar x_extent = 0.0;
+  Scalar y_extent = 0.0;
+
+  for (auto it = sub_atlas.begin(); it != sub_atlas.end(); it++) {
+    // This size was arbitrarily chosen to keep the textures from getting too
+    // wide. We could instead use a more generic rect packer but in the majority
+    // of cases the sample rects will be fairly close in size making this a good
+    // enough approximation.
+    if (x_offset >= 1000) {
+      y_offset = y_extent + 1;
+      x_offset = 0.0;
+    }
+
+    auto key = it->first;
+    auto transforms = it->second;
+
+    auto new_rect = Rect::MakeXYWH(x_offset, y_offset, key.rect.size.width,
+                                   key.rect.size.height);
+    auto sub_transform = Matrix::MakeTranslation(Vector2(x_offset, y_offset));
+
+    x_offset += (key.rect.size.width + 1.0);
```

Comment:
actually this logic is probably wrong if the sample rect doesn't have an integral width. I should probably just ceil the width here too.